### DISCUSSION TITLE
Specifying different incompatible dependencies for different versions of python

### DISCUSSION THREAD

**Discussion by Sbozzolo** [python-poetry/poetry#4796] [Q&A] [answered]
I develop and maintain a library with poetry. This package depends on numpy and I want it to be compatible with Python3.6-3.10, so I specifcy `python = "^3.6"`. Numpy added support for python 3.10 in version 1.21, which is incompatible with python3.6 (note that numpy is an example of a package with this problem and my question is general). I would like to have conditional dependency: if we are working with python 3.6+, then depend on numpy 1.18, if we are working with python3.10, install numpy 1.21. I tried something like
```
    numpy = [{version = "^1.18.5", python = ">=3.6, <3.10"},
             {version = "^1.21.4", python = ">=3.10"}]
```
But poetry complains that 
```
The current project's Python requirement (>=3.6,<4.0) is not compatible with some of the required packages Python requirement:
        - numpy requires Python >=3.7,<3.11, so it will not be satisfied for Python >=3.6,<3.7 || >=3.11,<4.0. 
```
Is it possible to do what I want to do?

With ` setup.py`, one could do something like:
```
    if sys.version.split(".")[1] < 10:
        install_requires.append("numpy==1.18.5")
    else:
        install_requires.append("numpy==1.21.4")
```

Thanks

**Comment by abn** [comment, marked as answer]
You can either updated your project to use `python = "^3.6, <3.11"` or update your numpy requirement to be:

```toml
numpy = [
    { version = "^1.18.5", python = ">=3.6, <3.10" },
    { version = "^1.21.4", python = ">=3.10, <3.11" },
]
```

The issue is by default the upper bound of `^3.6` is `4.0` (special case). This implies the project, unless explicitly specified supports all new 3.x versions of after 3.6. The changes above declares that either your project is only known good till 3.11 or that numpy will be installed only for versions `<3.11`. Personally the more correct solution is to specifiy your python support version upper bound.

**Reply by Sbozzolo** [reply]
Thanks, that worked!
